filetracker/servers/storage.py

```diff
@@ -0,0 +1,257 @@
+                If True, the compression is skipped, and file is written as-is.
+                Note that the current server implementation sends
+                'Content-Encoding' header anyway, mandating client to
+                decompress the file.
+            digest: SHA256 digest of the file before compression
+                If specified, the digest will not be computed again, saving
+                resources.
+        """
+        link_path = self._link_path(name)
+        if os.path.exists(link_path) and _file_version(link_path) > version:
+            return
+
+        file_lock = self._lock_file(os.path.join('links', name))
+
+        # Path to temporary file that may be created in some cases.
+        temp_file_path = None
+
+        if digest is None:
+            # Write data to temp file and calculate hash.
+            temp_file_fd, temp_file_path = tempfile.mkstemp()
+            temp_file = os.fdopen(temp_file_fd, 'wb')
+            _copy_stream(data, temp_file, size)
+            temp_file.close()
+
+            if compressed:
+                # If data was already compressed, we have to decompress it
+                # before calculating the digest.
+                raw_file_fd, raw_file_path = tempfile.mkstemp()
+                raw_file = os.fdopen(raw_file_fd, 'wb')
+
+                with gzip.open(temp_file_path, 'rb') as compressed_file:
+                    shutil.copyfileobj(compressed_file, raw_file)
+                raw_file.close()
+
+                digest = _file_digest(raw_file_path)
+                os.unlink(raw_file_path)
+            else:
+                digest = _file_digest(temp_file_path)
+
+        blob_path = self._blob_path(digest)
+
+        prefix = digest[0:2]
+        db_lock = self._lock_file(os.path.join('db', prefix))
+
+        try:
+            link_count = int(self.db[name.encode('utf8')])
+        except KeyError:
+            link_count = 0
+
+        self.db[name.encode('utf8')] = str(link_count + 1).encode('utf8')
+
+        if link_count == 0:
+            # Create a new blob.
+            _create_file_dirs(blob_path)
+            if compressed:
+                if temp_file_path:
+                    os.rename(temp_file_path, blob_path)
+                else:
+                    with open(blob_path, 'wb') as blob:
+                        _copy_stream(data, blob, size)
+            else:
+                if temp_file_path:
+                    with open(temp_file_path, 'rb') as raw,\
+                            gzip.open(blob_path, 'wb') as blob:
+                        shutil.copyfileobj(raw, blob)
+                else:
+                    with gzip.open(blob_path, 'wb') as blob:
+                        _copy_stream(data, blob, size)
+
+        if temp_file_path and os.path.exists(temp_file_path):
+            os.unlink(temp_file_path)
+            
+        self._unlock_file(db_lock)
+
+        if os.path.exists(link_path):
+            # Lend the link lock to delete().
```

Comment:
db_lock też możnaby pożyczyć, bo delete będzie musiało zmniejszyć licznik wystąpień bloba (ale to ja już mogę w swojej zmianie zmienić)

Reply:
Aaa racja. Błędnie założyłem, że wysyłany i istniejący plik są takie same.